pkg/pprof/fix_go_heap_truncated.go

```diff
@@ -0,0 +1,266 @@
+package pprof
+
+import (
+	"bytes"
+	"reflect"
+	"sort"
+	"unsafe"
+
+	"golang.org/x/exp/slices"
+
+	profilev1 "github.com/grafana/pyroscope/api/gen/proto/go/google/v1"
+)
+
+const (
+	minGroupSize = 2
+
+	tokens    = 8
+	tokenLen  = 16
+	suffixLen = tokens + tokenLen
+
+	tokenBytesLen  = tokenLen * 8
+	suffixBytesLen = suffixLen * 8
+)
+
+// MayHaveGoHeapTruncatedStacktraces reports whether there are
+// any chances that the profile may have truncated stack traces.
+func MayHaveGoHeapTruncatedStacktraces(p *profilev1.Profile) bool {
+	if !hasGoHeapSampleTypes(p) {
+		return false
+	}
+	// Some truncated stacks have depth less than the depth limit (32).
+	const minDepth = 28
+	for _, s := range p.Sample {
+		if len(s.LocationId) >= minDepth {
+			return true
+		}
+	}
+	return false
+}
+
+func hasGoHeapSampleTypes(p *profilev1.Profile) bool {
+	for _, st := range p.SampleType {
+		switch p.StringTable[st.Type] {
+		case
+			"alloc_objects",
+			"alloc_space",
+			"inuse_objects",
+			"inuse_space":
+			return true
+		}
+	}
+	return false
+}
+
+// RepairGoHeapTruncatedStacktraces repairs truncated stack traces
+// in Go heap profiles.
+//
+// Go heap profile has a depth limit of 32 frames, which often
+// renders profiles unreadable, and also increases cardinality
+// of stack traces.
+//
+// The function guesses truncated frames based on neighbors and
+// repairs stack traces if there are high chances that this
+// part is present in the profile. The heuristic is as follows:
+//
+// For each stack trace S taller than 24 frames: if there is another
+// stack trace R taller than 24 frames that overlaps with the given
+// one by at least 16 frames in a row from the top, and has frames
```

Comment:
why 16 and not less ?

Reply:
I think we should test in dev with different of that configuration just to see the impact.

Reply:
Less overlap size implies more tokens. The more tokens, the more checks we need to perform, the procedure becomes more expensive. In the meantime, with a shorter overlap, the chance that we will "repair" a wrong stack is higher. Therefore I'm not sure if it's worth it.

I experimented with various values and can say that the current ones work well. Thus, I would prefer to deploy this version and optimize it if we deem it necessary.